### ISSUE TITLE
Add Comprehensive Distributed Training Support

### ISSUE THREAD

**Issue by athul-22** [tracel-ai/burn#3413]
## Problem Statement
Currently, Burn lacks comprehensive support for distributed training across multiple devices and nodes, which is essential for training large-scale deep learning models efficiently. While Burn has excellent multi-backend support and thread-safe building blocks, there's no native distributed training infrastructure comparable to PyTorch's `torch.distributed` or TensorFlow's distributed strategies.

## Current State
- Burn has thread-safe modules that can be sent between threads
- Individual backends support GPU acceleration
- No native support for:
  - Multi-node training coordination
  - Gradient synchronization across devices
  - Data parallel training strategies
  - Model parallel training for large models

## Proposed Solution
Implement a comprehensive distributed training system that includes:

### Core Components
1. **Distributed Backend Abstraction**
   - Communication primitives (AllReduce, AllGather, Broadcast, etc.)
   - Support for different communication backends (NCCL, Gloo, MPI)
   - Cross-platform compatibility

2. **Training Strategies**
   - Data Parallel: Distribute batches across multiple devices
   - Model Parallel: Split large models across devices
   - Pipeline Parallel: Stage-wise model execution
   - Hybrid strategies for complex scenarios

3. **Synchronization Mechanisms**
   - Gradient accumulation and averaging
   - Parameter synchronization
   - Asynchronous and synchronous training modes

4. **Launch and Coordination**
   - Process group management
   - Device topology detection
   - Fault tolerance and recovery

### Implementation Details
```rust
// Example API design
use burn::distributed::{DistributedConfig, DistributedLearner};

let config = DistributedConfig::new()
    .with_strategy(DataParallel)
    .with_world_size(4)
    .with_backend(NcclBackend::new());

let learner = DistributedLearner::new(model, config)
    .build(&device);

// Training loop with automatic gradient synchronization
learner.train(dataloader);
```

## Benefits
- **Scalability**: Enable training of larger models and datasets
- **Performance**: Leverage multiple GPUs/nodes for faster training
- **Ecosystem Growth**: Make Burn competitive with PyTorch/TensorFlow for large-scale ML
- **Research Enablement**: Support cutting-edge research requiring distributed computation

## Technical Considerations
- Leverage Burn's existing thread-safe architecture
- Maintain backend agnosticism (work with WGPU, CUDA, etc.)
- Ensure compatibility with existing Burn APIs
- Optimize for both speed and memory efficiency

## Success Criteria
- [ ] Basic data parallel training across multiple GPUs
- [ ] Multi-node training support
- [ ] Integration with existing Burn backends
- [ ] Comprehensive documentation and examples
- [ ] Performance benchmarks showing scaling efficiency
- [ ] Test suite covering various distributed scenarios

## Related Work
- PyTorch Distributed: `torch.distributed`
- TensorFlow Distributed: `tf.distribute.Strategy`
- Horovod: Framework-agnostic distributed training

This feature would significantly enhance Burn's capabilities for large-scale machine learning workloads and make it more competitive in the enterprise and research markets.

**Comment by laggui** [comment]
Hey thanks for your interest!

This is a duplicate of #3261, though your issue is a bit more detailed.. perhaps even too big in scope 😅 

There is some ongoing work on collective operations.

**Comment by antimora** [comment]
Closing as a duplicate of #3261, as noted above.
